openai/api_requestor.py

```diff
@@ -89,32 +89,33 @@ def _make_session() -> requests.Session:
     return s
 
 
-def parse_stream_helper(line: bytes):
+def parse_stream_helper(line: bytes) -> Optional[str]:
     if line:
         if line.strip() == b"data: [DONE]":
             # return here will cause GeneratorExit exception in urllib3
             # and it will close http connection with TCP Reset
             return None
-        if hasattr(line, "decode"):
-            line = line.decode("utf-8")
-        if line.startswith("data: "):
-            line = line[len("data: ") :]
-        return line
+        if line.startswith(b"data: "):
+            line = line[len(b"data: ") :]
+        return line.decode("utf-8")
     return None
 
 
-def parse_stream(rbody):
+def parse_stream(rbody: Iterator[bytes]) -> Iterator[str]:
     for line in rbody:
         _line = parse_stream_helper(line)
         if _line is not None:
             yield _line
 
 
 async def parse_stream_async(rbody: aiohttp.StreamReader):
-    async for line, _ in rbody.iter_chunks():
-        _line = parse_stream_helper(line)
-        if _line is not None:
-            yield _line
+    async for chunk in rbody:
+        # While the `StreamReader` iterator is meant to iterate over lines by default, it seems like it sometimes
+        # returns multiple lines at a time, so let's split the chunk by lines again.
+        for line in chunk.splitlines():
```

Comment:
Why do you prefer to change what the server sends here? iter_chunks returns exactly the chunks as received from the server

Reply:
When stress testing this I saw that `iter_chunks` would return multiple lines in a chunk maybe 1/3 of the time so I believe we still need to handle this edge case. We could go back to using `iter_chunks` rather than the regular async stream iterator though, you're right.